### PR TITLE
EE-1120: Slash unbonding delegators.

### PR DESCRIPTION
**Pull request by mpapierski**
Ref: https://casperlabs.atlassian.net/browse/EE-1120

Adds a clause to the slashing logic so delegators present in bids entry for given slashed validator are collected, and then also removed from unbonding purses for given slashed validator.

This also inlines `detail::quash_bid` to get the benefit of removing validator's delegators as part of single step per validator to be slashed.

### CODE REVIEW

**Review by henrytill — COMMENTED** [review]
I think we need new slash tests exercising this logic.

File: grpc/tests/src/test/regression/ee_1120.rs [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+            ARG_VALIDATOR_PUBLIC_KEYS => vec![
+                VALIDATOR_2
+            ]
+        },
+    )
+    .build();
+
+    builder.exec(slash_request_1).expect_success().commit();
+
+    // compare bids after slashing validator 2
```
Comment (henrytill) [inline grpc/tests/src/test/regression/ee_1120.rs:248]:
```suggestion
    // Compare bids after slashing validator 2
```

File: grpc/tests/src/test/regression/ee_1120.rs [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+    let bids_after: Bids = builder.get_value(auction, BIDS_KEY);
+    assert_ne!(bids_before, bids_after);
+    assert_eq!(bids_after.len(), 1);
+    assert!(!bids_after.contains_key(&VALIDATOR_2));
+
+    assert!(bids_after.contains_key(&VALIDATOR_1));
+    assert_eq!(bids_after[&VALIDATOR_1].delegators().len(), 2);
+
+    // NOTE: Validator2's bid on Validator1 wasn't slashed.
```
Comment (henrytill) [inline grpc/tests/src/test/regression/ee_1120.rs:257]:
```suggestion
    // validator 2's delegation bid on validator 1 was not slashed.
```

File: grpc/tests/src/test/regression/ee_1120.rs [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+    let validator_1_unbond_list_after = unbond_purses_after
+        .get(&VALIDATOR_1)
+        .expect("should have validator 1 entry");
+    assert_eq!(validator_1_unbond_list_after.len(), 2);
+    assert_eq!(
+        validator_1_unbond_list_after[0].unbonder_public_key(),
+        &NONVALIDATOR_1
+    );
+
+    // NOTE: As mentioned above Validator2's unbonding purse on Validator1 wasn't slashed
```
Comment (henrytill) [inline grpc/tests/src/test/regression/ee_1120.rs:277]:
```suggestion
    // validator 2's delegation unbond from validator 1 was not slashed
```

File: grpc/tests/src/test/regression/ee_1120.rs [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+    assert_eq!(
+        validator_1_unbond_list_after[1].unbonder_public_key(),
+        &VALIDATOR_2
+    );
+
+    // if a delegator gets slashed for one validator's behavior, he isn't also slashed for the other
+    // validators he delegates to
```
Comment (henrytill) [inline grpc/tests/src/test/regression/ee_1120.rs:284]:
```suggestion
    // delegator 1 had a delegation unbond slashed for validator 2's behavior.
    // delegator 1 still has an active delegation unbond from validator 2.
```

File: grpc/tests/src/test/regression/ee_1120.rs [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+    assert_eq!(
+        validator_1_unbond_list_after,
+        &validator_1_unbond_list_before
+    );
+
+    //
+    // slash validator1 to clear both bids and unbonding purses
```
Comment (henrytill) [inline grpc/tests/src/test/regression/ee_1120.rs:291]:
```suggestion
    // slash validator 1 to clear remaining bids and unbonding purses
```

File: grpc/tests/src/test/regression/ee_1120.rs [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+        .exec(delegate_exec_request_2)
+        .expect_success()
+        .commit();
+
+    builder
+        .exec(delegate_exec_request_3)
+        .expect_success()
+        .commit();
+
+    // sanity check of the system before undelegating
```
Comment (henrytill) [inline grpc/tests/src/test/regression/ee_1120.rs:148]:
```suggestion
    // Ensure that initial bid entries exist for validator 1 and validator 2
```

File: types/src/auction.rs [hunk 7]
```diff
@@ -263,18 +266,34 @@ pub trait Auction:
...
+        let mut burned_amount: U512 = U512::zero();
 
-        let mut unbonding_purses: UnbondingPurses = detail::get_unbonding_purses(self)?;
+        let mut bids = detail::get_bids(self)?;
+        let mut modified_validators = 0usize;
 
+        let mut unbonding_purses: UnbondingPurses = detail::get_unbonding_purses(self)?;
         let mut unbonding_purses_modified = false;
         for validator_public_key in validator_public_keys {
+            // Clean up bids for given validator and save up assigned delegators
```
Comment (henrytill) [inline types/src/auction.rs:277]:
```suggestion
            // Remove bid for given validator, saving its delegators
```

File: types/src/auction.rs [hunk 8]
```diff
@@ -263,18 +266,34 @@ pub trait Auction:
...
+                Some(bid) => {
+                    burned_amount += *bid.staked_amount();
+                    modified_validators += 1;
+
+                    bid.delegators().keys().copied().collect()
+                }
+                None => BTreeSet::new(),
+            };
+
+            // Clean up unbonding entry for given validator
```
Comment (henrytill) [inline types/src/auction.rs:288]:
```suggestion
            // Update unbonding entries for given validator
```

File: types/src/auction.rs [hunk 9]
```diff
@@ -263,18 +266,34 @@ pub trait Auction:
...
             if let Some(unbonding_list) = unbonding_purses.get_mut(&validator_public_key) {
                 let initial_length = unbonding_list.len();
 
                 unbonding_list.retain(|unbonding_purse| {
-                    // Only entries created by non-validators are retained
-                    let should_retain = !unbonding_purse.is_validator();
+                    // Only entries created by validator entries and delegators that were attached
+                    // to quashed bids of validator.
```
Comment (henrytill) [inline types/src/auction.rs:294]:
```suggestion
                    // Retain all entries where this validator is delegating to another validator.
```

(if my assumptions in the other comment are correct)

File: types/src/auction.rs [hunk 10]
```diff
@@ -300,6 +319,10 @@ pub trait Auction:
             detail::set_unbonding_purses(self, unbonding_purses)?;
         }
 
+        if modified_validators > 0 {
+            detail::set_bids(self, bids)?;
+        }
+
         // call reduce total supply
```
Comment (henrytill) [inline types/src/auction.rs:326]: Remove this line, it's redundant

**Review by henrytill — COMMENTED** [review]
Overall looking good.  A couple suggestions and questions.

With regards to the test, these may seem like nits, but I'd like to see more consistency with regards to comments and formatting.  These auction tests are being read closely by several people (possibly third parties), so it is important that they be as clean as possible.

File: types/src/auction.rs (hunk 10 above)
Comment (mpapierski) [reply types/src/auction.rs:325]: Did this

File: grpc/tests/src/test/regression/ee_1120.rs [hunk 11]
```diff
@@ -0,0 +1,311 @@
...
+const TRANSFER_AMOUNT: u64 = MINIMUM_ACCOUNT_CREATION_BALANCE;
+
+const ARG_AMOUNT: &str = "amount";
+
+const SYSTEM_ADDR: AccountHash = AccountHash::new([0u8; 32]);
+const VALIDATOR_1: PublicKey = PublicKey::Ed25519([3; 32]);
+const VALIDATOR_2: PublicKey = PublicKey::Ed25519([4; 32]);
+const NONVALIDATOR_1: PublicKey = PublicKey::Ed25519([5; 32]);
+static VALIDATOR_1_ADDR: Lazy<AccountHash> = Lazy::new(|| VALIDATOR_1.into());
+static VALIDATOR_2_ADDR: Lazy<AccountHash> = Lazy::new(|| VALIDATOR_2.into());
```
[the remark this answers is not in the document]
Comment (mpapierski) [reply grpc/tests/src/test/regression/ee_1120.rs:39]: Done that

File: types/src/auction.rs [hunk 12]
```diff
@@ -263,18 +266,33 @@ pub trait Auction:
...
+                    burned_amount += *bid.staked_amount();
+                    bids_modified = true;
+
+                    bid.delegators().keys().copied().collect()
+                }
+                None => BTreeSet::new(),
+            };
+
+            // Update unbonding entries for given validator
             if let Some(unbonding_list) = unbonding_purses.get_mut(&validator_public_key) {
```
Comment (henrytill) [inline types/src/auction.rs:289]: Discussed offline: need to change this to `remove`.

**Review by henrytill — CHANGES_REQUESTED** [review]

File: types/src/auction.rs [hunk 13]
```diff
@@ -263,18 +266,34 @@ pub trait Auction:
...
                 let initial_length = unbonding_list.len();
 
                 unbonding_list.retain(|unbonding_purse| {
-                    // Only entries created by non-validators are retained
-                    let should_retain = !unbonding_purse.is_validator();
+                    // Only entries created by validator entries and delegators that were attached
+                    // to quashed bids of validator.
+                    let should_retain = !unbonding_purse.is_validator()
+                        && !orphaned_delegators.contains(unbonding_purse.unbonder_public_key());
```
[the remark this answers is not in the document]
Comment (henrytill) [reply types/src/auction.rs:295]: Addressed offline, my understanding did not reflect changes to unbonding purse structure.

File: types/src/auction.rs [hunk 14]
```diff
@@ -275,42 +272,18 @@ pub trait Auction:
...
-                    // Cleans up empty map entries to preserve global state.
-                    unbonding_purses.remove(&validator_public_key).unwrap();
-                    unbonding_purses_modified = true;
-                }
+            if let Some(unbonding_list) = unbonding_purses.remove(&validator_public_key) {
+                burned_amount += unbonding_list
+                    .into_iter()
+                    .map(|unbonding_purse| *unbonding_purse.amount())
+                    .sum();
+                unbonding_purses_modified = true;
```
Comment (henrytill) [inline types/src/auction.rs:286]: The symmetry between how we now handle `bids` and `unbonding_purses` is very reassuring!

**Review by henrytill — APPROVED** [review]

**Pull request merged**